Mathlib/CategoryTheory/Adjunction/DiscreteObjects.lean

```diff
@@ -0,0 +1,104 @@
+/-
+Copyright (c) 2024 Dagur Asgeirsson. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Dagur Asgeirsson
+-/
+import Mathlib.CategoryTheory.Adjunction.FullyFaithful
+import Mathlib.CategoryTheory.Adjunction.Unique
+/-!
+
+# Discrete objects
+
+This file defines the property of a functor of having discrete objects. This simply means that it
+has a fully faithful left adjoint.
+
+We also define a predicate `IsDiscrete` on objects saying that the counit of the adjunction applied
+to `X` is an isomorphism. This is equivalent to being in the essential image of the left adjoint
+(see `isDiscrete_iff_mem_essImage`).
+-/
+
+open CategoryTheory
+
+variable {C D : Type*} [Category C] [Category D] (U : C ⥤ D)
+
+namespace CategoryTheory.Functor
+
+/-- A functor is said to have discrete objects if it has a fully faithful left adjoint. -/
+class HasDiscreteObjects : Prop where
```

Comment:
I believe that we could have the adjunction part of the class similarly as it was done in `Adjunction.Reflective`. Also, if the typeclass was attached to the left adjoint rather that to the right adjoint, it seems this would be the notion of "coreflective" functor, which is dual to reflective functors. Then, I think it would make sense to dualize (part of?) the results from `Adjunction.Reflective` into a new file `Adjunction.Coreflective` instead. This may also remove the need for #13923.

Reply:
I dualized some of the API from `Adjunction.Reflective` in #13948. Then I ended up not needing it for discrete sheaves, which I've defined at #13947 without using this ` DiscreteObjects` API. 

One can easily prove `Coreflective (constantSheaf J A)` if it is fully faithful and the site has a terminal object. However, I want to talk about discreteness with respect to a specific terminal object (this is because in `CompHaus` and friends, it is useful to use the terminal object `CompHaus.of PUnit` instead of the arbitrary one given by `HasTerminal`) and I do this successfully in #13497 (needs some cleanup before it's ready for review).

So I think I'll just close all these `DiscreteObjects` PRs, one should use the `Coreflective` API instead if this is desired.